Use Python threading to avoid creating temp files

6-6-6 opened the pull request:
This is an effort to avoid creating temp files under `/tmp` folder.

It uses Python's threading and queueing library to receive and preprocess the output of `powermetrics`.

The main change is presented in commit 1daf3add70f113a8f1e97f0d410231c7b57f0cec
The other 2 commits are for code cleanups.

The whole process is described as follows:
1. `powermetrics` is configured to output to `stdout`
2. a LIFO queue is created for communicating between threads
3. the output will be passed to another thread and processed by `enqueue_powermetrics()`, which will
    - convert the lines of `stdout` into separate plists (by the magic `b'\x00'` character)
    - each plist will be added to the LIFO queue
4. instead of reading the whole temp file, `parse_powermetrics()` will get the latest plist from the LIFO queue, and parse the plist as usual

This PR will hopefully resolve issues #18, #17 and #35.
Also #6 may also be resolved (without restarting `powermetrics` multiple times).

Lunar-ZZ committed:
Use another thread to combine the output of powertrics

making asitop not need to write file on the filesystem

Signed-off-by: Zhang Zongyu <zongyu@novazy.net>

Lunar-ZZ committed:
cleanup codes about creating temp files

Signed-off-by: zongyu <zongyu@novazy.net>

Lunar-ZZ committed:
remove code about restarting `powermetric` process

since temp files do not exists anymore,
the memory consumption problem no longer exists.

Signed-off-by: zongyu <zongyu@novazy.net>

buliaoyin commented:
Seems nice, I'll try to check if the memory issue had been resolved.

buliaoyin commented:
> Seems nice, I'll try to check if the memory issue had been resolved.

After 6 hours, the python process memory was still only around 30MB, memory issue has gone.

Lunar-ZZ committed:
suggested by "Codacy Static Code Analysis"

Signed-off-by: Zhang Zongyu <zongyu@novazy.net>

Lunar-ZZ committed:
asitop/asitop.py: no need to sleep during the main loop

Waiting for the queue to be filled by utils.enqueue_powermetrics()
  implies that we need to sleep for args.interval seconds,
  so we do not need extra lines to sleep between `powermetrcis` outputs.

Signed-off-by: Zhang Zongyu <zongyu@novazy.net>

6-6-6 commented:
Hi, I have just added two more commits.

https://github.com/tlkh/asitop/pull/39/commits/1acead9e3d0216324f5c62b82a04fb5c3a03a31f is to meet the requirement of "Codacy Static Code Analysis".

https://github.com/tlkh/asitop/pull/39/commits/5b6b58c03f4b02523f8e55fd98e5809a6a59093d is to remove the redundant `time.sleep(args.interval)` in `asitop.asitop.py`,
and the reason is explained in the commit message.

6-6-6 committed:
Merge branch 'tlkh:main' into pipe

Lunar-ZZ committed:
Merge branch 'main' into pipe

Lunar-ZZ committed:
Merge branch 'main' into pipe

Vlor999 commented:
Hi everyone, since this project seems to be unmaintained and has accumulating bugs with recent macOS updates, I've created an actively maintained fork called mxtop. It includes pending community fixes and is fully working on the latest Apple Silicon chips.
You can install it right now using: pip install mxtop The new repository is here: https://github.com/Vlor999/mxtop
Feel free to open issues or PRs there!
